File: csharp/Slimmed/PublishingHouse.Domain/Books/BookEvent.cs
using PublishingHouse.Books.Entities;
using PublishingHouse.Core;
using PublishingHouse.Core.Events;
using PublishingHouse.Core.ValueObjects;

namespace PublishingHouse.Books;

public abstract record BookEvent: IDomainEvent
{
    public record DraftCreated(
        BookId BookId,
        Title Title,
        Author Author,
        Publisher Publisher,
        PositiveInt Edition,
        Genre? Genre
    ): BookEvent;

    public record ChapterAdded(
        BookId BookId,
        Chapter Chapter
    ): BookEvent;

    public record FormatAdded(
        BookId BookId,
        Format Format
    ): BookEvent;

    public record FormatRemoved(
        BookId BookId,
        Format Format
    ): BookEvent;

    public record TranslationAdded(
        BookId BookId,
        Translation Translation
    ): BookEvent;

    public record TranslationRemoved(
        BookId BookId,
        Translation Translation
    ): BookEvent;

    public record ReviewerAdded(
        BookId BookId,
        Reviewer Reviewer
    ): BookEvent;

    public record MovedToEditing(
        BookId BookId,
        Genre Genre
    ): BookEvent;

    public record Approved(
        BookId BookId,
        CommitteeApproval CommitteeApproval
    ): BookEvent;

    public record ISBNSet(
        BookId BookId,
        ISBN ISBN
    ): BookEvent;

    public record MovedToPrinting
    (
        BookId BookId
    ): BookEvent;

    public record Published(
        BookId BookId
    ): BookEvent;

    public record MovedToOutOfPrint
    (
        BookId BookId
    ): BookEvent;

    private BookEvent() { }
}

public abstract record BookExternalEvent: IDomainEvent
{
    public record Published(
        BookId BookId,
        ISBN ISBN,
        Title Title,
        AuthorId AuthorId
    ): BookExternalEvent;

    private BookExternalEvent() { }
}

File: node.js/src/slimmed/domain/books/bookEvent.ts
import { PositiveNumber } from '#core/typing';
import { DomainEvent } from '../../infrastructure/events';
import {
  Author,
  AuthorId,
  BookId,
  Chapter,
  CommitteeApproval,
  Format,
  Genre,
  ISBN,
  Publisher,
  Reviewer,
  Title,
  Translation,
} from './entities';

export type DraftCreated = DomainEvent<
  'DraftCreated',
  {
    bookId: BookId;
    title: Title;
    author: Author;
    publisher: Publisher;
    edition: PositiveNumber;
    genre: Genre | null;
  }
>;

export type ChapterAdded = DomainEvent<
  'ChapterAdded',
  {
    bookId: BookId;
    chapter: Chapter;
  }
>;

export type FormatAdded = DomainEvent<
  'FormatAdded',
  {
    bookId: BookId;
    format: Format;
  }
>;

export type FormatRemoved = DomainEvent<
  'FormatRemoved',
  {
    bookId: BookId;
    format: Format;
  }
>;

export type TranslationAdded = DomainEvent<
  'TranslationAdded',
  {
    bookId: BookId;
    translation: Translation;
  }
>;

export type TranslationRemoved = DomainEvent<
  'TranslationRemoved',
  {
    bookId: BookId;
    translation: Translation;
  }
>;

export type ReviewerAdded = DomainEvent<
  'ReviewerAdded',
  {
    bookId: BookId;
    reviewer: Reviewer;
  }
>;

export type MovedToEditing = DomainEvent<
  'MovedToEditing',
  {
    bookId: BookId;
    genre: Genre | null;
  }
>;

export type Approved = DomainEvent<
  'Approved',
  {
    bookId: BookId;
    committeeApproval: CommitteeApproval;
  }
>;

export type ISBNSet = DomainEvent<
  'ISBNSet',
  {
    bookId: BookId;
    isbn: ISBN;
  }
>;

export type MovedToPrinting = DomainEvent<
  'MovedToPrinting',
  {
    bookId: BookId;
  }
>;

export type Published = DomainEvent<
  'Published',
  {
    bookId: BookId;
  }
>;

export type MovedToOutOfPrint = DomainEvent<
  'MovedToOutOfPrint',
  {
    bookId: BookId;
  }
>;

export type DraftEvent = DraftCreated | ChapterAdded;

export type UnderEditingEvent =
  | MovedToEditing
  | TranslationAdded
  | TranslationRemoved
  | FormatAdded
  | FormatRemoved
  | ReviewerAdded
  | Approved
  | ISBNSet;

export type InPrintEvent = MovedToPrinting;

export type PublishedEvent = Published;

export type OutOfPrintEvent = MovedToOutOfPrint;

export type BookEvent =
  | DraftEvent
  | UnderEditingEvent
  | InPrintEvent
  | PublishedEvent
  | OutOfPrintEvent;

export type PublishedExternal = DomainEvent<
  'Published',
  {
    bookId: BookId;
    isbn: ISBN;
    title: Title;
    authorId: AuthorId;
  }
>;

export type BookExternalEvent =
  | DraftEvent
  | UnderEditingEvent
  | PublishedExternal
  | OutOfPrintEvent;
